src/isc-dhcp/patch/0009-Support-for-dual-tor-scenario.patch

```diff
@@ -0,0 +1,308 @@
+@@ -862,6 +884,8 @@ main(int argc, char **argv) {
+ 
+ 	/* In fact dispatch() never returns. */
+ 	free_interface_alias_map();
++	free_downstream_intfs();
++
+ 	return (0);
+ }
+ 
+@@ -885,25 +909,50 @@ do_relay4(struct interface_info *ip, struct dhcp_packet *packet,
+ 		return;
+ 	}
+ 
+-	/* Find the interface that corresponds to the giaddr
+-	   in the packet. */
+-	if (packet->giaddr.s_addr) {
+-		for (out = interfaces; out; out = out->next) {
+-			int i;
++	if (enable_support_for_dual_tor) {
++		if (packet->yiaddr.s_addr) {
++			out = NULL;
+ 
+-			for (i = 0 ; i < out->address_count ; i++ ) {
+-				if (out->addresses[i].s_addr ==
+-				    packet->giaddr.s_addr) {
+-					i = -1;
++			for (struct downstream_intf_list *cdi = downstream_intfs; cdi; cdi = cdi->next) {
++				int i = 0;
++				out = cdi->interface;
++
++				for (i = 0 ; i < out->address_count ; i++ ) {
++					if ((out->addresses[i].s_addr & out->netmasks[i].s_addr) == (packet->yiaddr.s_addr & out->netmasks[i].s_addr)) {
++						i = -1;
++						break;
++					}
++				}
++
++				if (i == -1) {
+ 					break;
++				} else {
++					out = NULL;
+ 				}
+ 			}
+-
+-			if (i == -1)
+-				break;
++		} else {
++			out = NULL;
+ 		}
+ 	} else {
+-		out = NULL;
++		/* Find the interface that corresponds to the giaddr in the packet. */
++		if (packet->giaddr.s_addr) {
++			for (out = interfaces; out; out = out->next) {
++				int i;
++
++				for (i = 0 ; i < out->address_count ; i++ ) {
++					if (out->addresses[i].s_addr ==
++						packet->giaddr.s_addr) {
++						i = -1;
++						break;
++					}
++				}
++
++				if (i == -1)
++					break;
++			}
++		} else {
++			out = NULL;
++		}
+ 	}
+ 
+ 	/* If it's a bootreply, forward it to the client. */
+@@ -913,6 +962,10 @@ do_relay4(struct interface_info *ip, struct dhcp_packet *packet,
+ 			return;
+ 		}
+ 
++		/* This bootreply does not belong to the current vlan. */
++		if (enable_support_for_dual_tor && !out)
++			return;
++
+ 		if (!(packet->flags & htons(BOOTP_BROADCAST)) &&
+ 			can_unicast_without_arp(out)) {
+ 			to.sin_addr = packet->yiaddr;
+@@ -945,9 +998,13 @@ do_relay4(struct interface_info *ip, struct dhcp_packet *packet,
+ 			return;
+ 
+ 		if (!out) {
+-			log_error("Packet to bogus giaddr %s.\n",
+-			      inet_ntoa(packet->giaddr));
+-			++bogus_giaddr_drops;
++			if (!enable_support_for_dual_tor) {
++				log_error("Packet to bogus giaddr %s.\n", inet_ntoa(packet->giaddr));
++				++bogus_giaddr_drops;
++			} else {
++				log_error("Packet to bogus yiaddr %s.\n", inet_ntoa(packet->yiaddr));
++				++bogus_yiaddr_drops;
++			}
+ 			return;
+ 		}
+ 
+@@ -987,8 +1044,14 @@ do_relay4(struct interface_info *ip, struct dhcp_packet *packet,
+ 	   forward the response to the correct net.    If it's already
+ 	   set, the response will be sent directly to the relay agent
+ 	   that set giaddr, so we won't see it. */
+-	if (!packet->giaddr.s_addr)
+-		packet->giaddr = ip->addresses[0];
++	if (!packet->giaddr.s_addr) {
```

Comment:
Is this code snippet required? looks like the `-U` option will update the both `giaddr` and Option 82 sub-option 5, right?

Reply:
this if clause should not be triggered at all, but I follow the existing code logic in case we miss some cases.

Reply:
It is still not required as this is applicable for the generic case. With `-U` option, the giaddr is updated elsewhere per the following excerpt:

```
		/* RFC3527: Use the inbound packet's interface address in
		 * the link selection suboption and set the outbound giaddr
		 * to the uplink address. */
		if (adding_link_select) {
			*sp++ = RAI_LINK_SELECT;
			*sp++ = 4u;
			memcpy(sp, &giaddr.s_addr, 4);
			sp += 4;
			packet->giaddr = uplink->addresses[0];
			log_debug ("Adding link selection suboption"
				   " with addr: %s", inet_ntoa(giaddr));
		}
```

Reply:
updated